### ISSUE TITLE
Multiple output intervals?

### ISSUE THREAD

**Issue by alexmitchell** [TerrainBento/terrainbento#154]
Great project! I've gotten a custom model up and running without too much trouble. The docs, notebooks, and 2019 paper are very helpful. I have a question/feature request for model output: is there a way to output at two (or more) intervals?

Terrainbento automatically writes netcdf  files every `output_interval`, which is often much longer than the model time step. No issues with that. However, I would like to also output total sediment yield from every model step. After the model runs I would like to make a plot of sediment yield changing over time. (Eventually distinguishing yields from different process components. But that is an issue for another day.)

I've gone through the docs, jupyter notebooks, and source code. I know I can make a custom `output_writers` function/class which calculates the desired value and appends it to a file. However, the `write_output` function in `ErosionModel` also writes a netcdf file when it calls `run_one_step` for the output writers. So setting the `output_interval` to every model step will also write a netcdf file every step. Output classes can remember previous values, but only accesses the model data when its `run_one_step` is called (correct?). If I use a longer `output_interval`, the output class would miss any model steps run between the `output_interval`.

In summary, it seems that terrainbento will let me either: 1) have reasonable netcdf output frequency but low temporal resolution for the sediment yields or 2) have high temporal resolution on the sediment yields but waste resources saving many unwanted netcdf files and wait approximately one forever to generate the output gif/mp4.

How doable do you think it would be to modify terrainbento to allow each output writer (including the netcdf one) to have it's own output interval and ErosionModel to have a queue of what needs to output when?

Thank you.

**Comment by kbarnhart** [comment]
@alexmitchell I think your assessment of current capabilities is accurate. 

I am pretty sure that this could be accomplished by modifying only the base class erosion model (in `terrainbento/base_class/erosion_model.py`).

This isn't a feature that I can support. But I'd be happy to provide feedback/guidance if you are interested in making a PR implementing it. 

I think a general solution would permit netcdf output and each output writer to be called at an independent interval (fixed interval size or a series defined in say, a list). An intermediate goal might be that netcdf output is created on one set of intervals, and all output writers are called on another set of intervals. The function call `run_output_writers` would get taken out of `write_output` and live in `run` to be called at whatever interval is specified. 

Those intervals and what gets called when probably need to be identified in the init. Then the `run` function would need to be modified so that it stops at each interval, and then calls the correct set of routines to output. 

One thing that might be tricky is making sure that the output filenames all make sense (right now they are based on self.iteration, that might not make the most sense with multiple output time intervals).

If temporary space is not a barrier, and profiling indicates that not much time is spent writing the netcdfs, another option would be to output netcdfs at the high temporal frequency, then delete all but every N of them before doing analyses of the entire spatial domain (e.g., making animations). 

Let me know if you have questions I haven't addressed. If I have any other thoughts on how this might be addressed I'll post them here.

**Comment by alexmitchell** [comment]
Thank you for the detailed feedback and ideas. I'll give it a try after I finish my current set of tasks.

I will aim towards doing the more general solution. Doesn't seem like that much extra work from the intermediate goal but adds quite a bit of flexibility and would keep the model code cleaner. My tentative thoughts are to write a base output class which interfaces with ErosionModel. Specific output types inherit the base class and define their own output functions and behavior. Either the erosion model would compile a sorted list of output times for the whole model duration in its constructor or the erosion model queries each output class for their next output time during the model run. The latter option would allow for dynamic output intervals based on current conditions, but I don't know how useful that would be. I would guess that they are similar in difficultly programming and performance, so I'm leaning towards the latter option to give users more control should they want it.

I think this OO design would mesh better with the terrainbento/landlab modular format and allow people to generate custom output more easily too. (To which Randall Munroe responds with: The General Problem https://xkcd.com/974/ ) Naming output would be simplified because the user can provide a unique/descriptive name base when initializing an output class ( iteration or model_time can be appended).

I will close the issue for now, but may reopen it if I run in to trouble or have questions on best practices for contributing. It will be my first time adding to a major collaborative project with quality standards.

**Comment by alexmitchell** [comment]
On second thought, it seems I should leave the issue open until it is completed.

**Comment by kbarnhart** [comment]
@alexmitchell I like your approach so far. Let me know if it would be helpful to chat further as you explore this. My only initial thought on your comments is that right now outputwriters can be either classes or functions... and the base class only addresses the class  version. 

```
output_writers : dictionary of output writers.
            Classes or functions used to write incremental output (e.g. make a
            diagnostic plot). These should be passed in a dictionary with two
            keys: "class" and "function". The value associated with each of
            these should be a list containing the uninstantiated output
            writers. See the Jupyter notebook examples for more details.
```

but perhaps requesting outputwriters to be specified as list of dicts, in which each entry has the form:
```python
{"type: "class", 
 "instance", OutputWriterClass, 
 "outstep": 100 # this could be a float or list of timepoints?
}
```

Then in the init the cross-all-outputs set of output times could be identified. 

I think it is worth thinking about whether a solution can be made that is not backwards compatibility breaking... but I think this would be a substantial enough feature that if compatibility needs to be broken to implement it well, then that is just what should happen.

**Comment by alexmitchell** [comment]
Right. I forgot about that bit. I'll keep backwards compatibility in mind. I will have to think about it some more and get back to you.
